### ISSUE TITLE
Regression in price lookup since version 1.20

### ISSUE THREAD

**Issue by apauley** [simonmichael/hledger#1439]
As of hledger 1.20 most of my reports which converts to a different commodity causes hledger to hang while using lots of CPU.

I've done a git bisect which reported that this behaviour was introduced in commit https://github.com/simonmichael/hledger/commit/221a6d900199448b9670aa3d823e665f333b59fc (related to #1402).

In my minimal journal (below) I have a transaction in a base commodity, and a price for a target commodity. And then there are some seemingly unrelated `$` prices with the base commodity and others - but when these are present it introduces a severe slowdown in any report that attempts to convert to `TARGET`. Adding some more of the commented-out prices increases the slowdown significantly.

I've tested this on both OS X and Linux, and the results are similar.

My test journal:
```hledger
2020-12-31 Buy Stuff
  expenses:Misc           BASE 100
  assets:Cash

P 2020-12-31 TARGET BASE 2

P 2020-12-31 $ BASE 3

P 2020-12-31 A $10
P 2020-12-31 B $20
P 2020-12-31 C $30
P 2020-12-31 D $40
P 2020-12-31 E $50
P 2020-12-31 F $60
P 2020-12-31 G $70
P 2020-12-31 H $80
P 2020-12-31 I $90
P 2020-12-31 J $100
; P 2020-12-31 K $120
; P 2020-12-31 L $130
; P 2020-12-31 M $140
```

First a good run using a version of hledger built using the commit just before the offending commit:
```bash
$ time ~/.local/bin/hledger-good-13e3e76 -f ./test.journal balance --exchange=TARGET
           TARGET-50  assets:Cash
            TARGET50  expenses:Misc
--------------------
                   0

real	0m0,027s
user	0m0,020s
sys	0m0,005s
```

When running it with the newer version of hledger I see a severe slowdown:
```bash
$ time ~/.local/bin/hledger-bad-221a6d9 -f ./test.journal balance --exchange=TARGET
           TARGET-50  assets:Cash
            TARGET50  expenses:Misc
--------------------
                   0

real	0m21,287s
user	0m21,217s
sys	0m0,055s
```

Adding just one more `$` price (e.g. for commodity `K`) further slows this balance down to about 4 minutes on my machine. Adding`L` and `M` slows it down to a point where I didn't see it exit at all.

Interesting to note:
 - The `$`-sign seems to be somehow significant. The issue doesn't occur when changing `$` to any other symbol.
 - Removing the price between `$` and `BASE` removes the slowdown for the `TARGET` conversion, regardless of how many other `$` prices there are - `P 2020-12-31 $ BASE 3`.

**Comment by simonmichael** [comment]
@apauley, thank you very much. I suspect an infinite loop in Hledger.Data.Valuation.findPath, but I think I'm out of time for 2020. Better behaviour in 2021!

**Comment by simonmichael** [comment]
This is fixed in master:

> Searching for prices during valuation no longer now properly excludes
price loops, avoiding near infinite looping with certain
configurations of market prices. Also we now always use a direct price
when available, rather than searching unnecessarily.

> Price searching progress info, useful for troubleshooting, is now
displayed with --debug=2.

> There could still be some corner cases we don't handle correctly. We
now give up with an error message if the searched price chains get too
long (> 1000). More importantly, we should also give up if the search
iterates too many times, but this is not done yet.

**Comment by simonmichael** [comment]
Reopening as a reminder to give up after N iterations, to document this limit, and to release.

**Comment by apauley** [comment]
Thanks @simonmichael! I tried it out and the reports no longer hangs 😃 

The reports now return different values than before, but I guess that's to be expected after the recent price lookup changes.

**Comment by simonmichael** [comment]
That is not expected! Example please :)

**Comment by apauley** [comment]
OK, I played around a bit with some reports. I don't have a minimal journal unfortunately, I'll need to dig some more in order to create one.

It seems that there's no difference between 13e3e76 and the latest master (73678393b1ec9ea414d798ade9da6e5666c079c2) as with the previous bisect I did.
But there is a difference between 1.19 and 13e3e76.

I can trigger the difference by specifying both `--exchange` and an interval, eg `--yearly`.

My base good output shows that hledger 1.19 displays the same values with and without `--yearly`, for a report that is restricted to just one year:
```
$ ~/.local/bin/hledger-1.19 -f ./all-years.journal --period 20xx incomestatement --depth 3 --exchange=Z expenses:Test
Income Statement 20xx, current value

                 ||             20xx
=================++==================
 Revenues        ||
-----------------++------------------
-----------------++------------------
                 ||
=================++==================
 Expenses        ||
-----------------++------------------
 expenses:Test:A ||           Z34.07
 expenses:Test:B || Z-113.00
 expenses:Test:C ||          Z142.09
-----------------++------------------
                 ||           Z63.16
=================++==================
 Net:            ||  Z-63.16


$ ~/.local/bin/hledger-1.19 -f ./all-years.journal --period 20xx incomestatement --depth 3 --exchange=Z expenses:Test --yearly
Income Statement 20xx, valued at period ends

                 ||             20xx
=================++==================
 Revenues        ||
-----------------++------------------
-----------------++------------------
                 ||
=================++==================
 Expenses        ||
-----------------++------------------
 expenses:Test:A ||           Z34.07
 expenses:Test:B || Z-113.00
 expenses:Test:C ||          Z142.09
-----------------++------------------
                 ||           Z63.16
=================++==================
 Net:            ||  Z-63.16
```

hledger at commit 13e3e76 displays the same output as 1.19 without `--yearly`, but with `--yearly` the output is suddenly different.
The most striking difference is `expenses:Test:C` which goes from a positive amount to a negative amount:

```
~/.local/bin/hledger-13e3e76 -f ./all-years.journal --period 20xx incomestatement --depth 3 --exchange=Z expenses:Test
Income Statement 20xx, current value

                 ||     20xx
=================++==========
 Revenues        ||
-----------------++----------
-----------------++----------
                 ||
=================++==========
 Expenses        ||
-----------------++----------
 expenses:Test:A ||   Z34.07
 expenses:Test:B || Z-113.00
 expenses:Test:C ||  Z142.09
-----------------++----------
                 ||   Z63.16
=================++==========
 Net:            ||  Z-63.16

~/.local/bin/hledger-13e3e76 -f ./all-years.journal --period 20xx incomestatement --depth 3 --exchange=Z expenses:Test --yearly
Income Statement 20xx

                 ||     20xx
=================++==========
 Revenues        ||
-----------------++----------
-----------------++----------
                 ||
=================++==========
 Expenses        ||
-----------------++----------
 expenses:Test:A ||   Z32.69
 expenses:Test:B || Z-106.49
 expenses:Test:C || Z-203.90
-----------------++----------
                 || Z-277.70
=================++==========
 Net:            ||  Z277.70
```

**Comment by simonmichael** [comment]
That's interesting. I would guess d60c1f764 is responsible. 

In any case, I am expecting that the latest master (now refactored, d60c1f764) works perfectly - if you can find any repro showing otherwise, or a difference from 1.19, I'll definitely investigate.

**Comment by simonmichael** [comment]
PS with the latest code, use --debug=2 for additional clues.

**Comment by simonmichael** [comment]
Corrected my mis-paste: the latest master is 7c9303a15.

**Comment by apauley** [comment]
7c9303a does the same, there is a difference when adding `--yearly`:

```bash
$ ~/.local/bin/hledger-7c9303a -f ./all-years.journal --period 20xx incomestatement --depth 3 --exchange=Z expenses:Test
Income Statement 20xx, valued at period ends

                 ||     20xx
=================++==========
 Revenues        ||
-----------------++----------
-----------------++----------
                 ||
=================++==========
 Expenses        ||
-----------------++----------
 expenses:Test:A ||   Z34.07
 expenses:Test:B || Z-113.00
 expenses:Test:C ||  Z142.09
-----------------++----------
                 ||   Z63.16
=================++==========
 Net:            ||  Z-63.16


$ ~/.local/bin/hledger-7c9303a -f ./all-years.journal --period 20xx incomestatement --depth 3 --exchange=Z expenses:Test --yearly
Income Statement 20xx

                 ||     20xx
=================++==========
 Revenues        ||
-----------------++----------
-----------------++----------
                 ||
=================++==========
 Expenses        ||
-----------------++----------
 expenses:Test:A ||   Z32.69
 expenses:Test:B || Z-106.49
 expenses:Test:C || Z-203.90
-----------------++----------
                 || Z-277.70
=================++==========
 Net:            ||  Z277.70
```

**Comment by apauley** [comment]
I did another bisect, it points to 35a83fbd8c69ba228c6b9e9e62b72211b727f57e:

```
35a83fbd8c69ba228c6b9e9e62b72211b727f57e is the first bad commit
commit 35a83fbd8c69ba228c6b9e9e62b72211b727f57e
Author: Stephen Morgan <morgan.stephen99@gmail.com>
Date:   Mon Oct 26 23:47:50 2020 +1100

    lib: For MultiBalanceReport, report change in valuation rather than valuation of change.

 hledger-lib/Hledger/Reports/MultiBalanceReport.hs | 47 +++++++++++++++++------
 tests/journal/valuation.test                      | 40 +++++++++----------
 2 files changed, 55 insertions(+), 32 deletions(-)
```

I'll see if I can create a simplified journal file this week that reproduces the issue, and I'll have a look at the debug output you mentioned as well.

**Comment by simonmichael** [comment]
Thanks! I'll also do a bit more testing and generally catch up on valuation changes at the first opportunity.

**Comment by dominikpaulus** [comment]
I also stumbled over this. I created the following minimal example where above-mentioned commit changed the behavior in - for me - unexpected ways:

```
P 2018/01/31 € 1.14Fr
P 2021/01/04 € 1.08Fr

2019/05/31 test
   assets:cash  -2€
   expenses:test
```

hledger 1.20.2 produces the following output for me:
```
$ hledger balance -f ledger-test -M -X Fr --tree -b 2021/01/01 -e 2021/01/31 expenses
Period-end value changes in 2021-01:

               ||     Jan 
===============++=========
 expenses:test || -0.12Fr 
---------------++---------
               || -0.12Fr
```

whereas hledger 1.19 (with the same command line arguments) produces an empty output, as I've expected:
```
Balance changes in 2021-01, valued at period ends:

  || Jan 
==++=====
--++-----
  ||     
```

This matches my reading the commit message of 35a83fb. (hledger is now reporting the change in valuation of the 'expenses:test'-account, instead of the valuation of the changes to this account - and a change in exchange rate between € and Fr causes the valuation of the account, as measured in Fr, to change). Still, I think the change manifest itself in highly surprising (and imho, incorrect) output for multi-currency ledgers.

**Comment by apauley** [comment]
Thanks @dominikpaulus, the minimal journal I've been playing with only started to reproduce the issue when I added a second price.

```hledger
P 2020-01-01 COM $10
P 2021-01-01 COM $20

2020-01-01 Trx 1
    assets:A
    expenses:A:Year2020     COM 1

2021-01-01 Trx 2
    assets:A
    expenses:A:Year2021     COM 2
```

hledger 1.20 includes transactions that I think should be filtered out when given both an interval and an exchange with a change in price.

```bash
$ hledger-35a83fb -f test.journal is --begin 2021-01-01 --exchange=$
Income Statement 2021-01-01, current value

                     || 2021-01-01
=====================++============
 Revenues            ||
---------------------++------------
---------------------++------------
                     ||
=====================++============
 Expenses            ||
---------------------++------------
 expenses:A:Year2021 ||        $40
---------------------++------------
                     ||        $40
=====================++============
 Net:                ||       $-40

$ hledger-35a83fb -f test.journal is --begin 2021-01-01 --exchange=$ --daily
Income Statement 2021-01-01, valued at period ends

                     || 2021-01-01
=====================++============
 Revenues            ||
---------------------++------------
---------------------++------------
                     ||
=====================++============
 Expenses            ||
---------------------++------------
 expenses:A:Year2020 ||        $10
 expenses:A:Year2021 ||        $40
---------------------++------------
                     ||        $50
=====================++============
 Net:                ||       $-50
```

Before this hledger excluded the appropriate transactions in both cases:

```bash
$ hledger-524e23b -f test.journal is --begin 2021-01-01 --exchange=$
Income Statement 2021-01-01, current value

                     || 2021-01-01
=====================++============
 Revenues            ||
---------------------++------------
---------------------++------------
                     ||
=====================++============
 Expenses            ||
---------------------++------------
 expenses:A:Year2021 ||        $40
---------------------++------------
                     ||        $40
=====================++============
 Net:                ||       $-40

$ hledger-524e23b -f test.journal is --begin 2021-01-01 --exchange=$ --daily
Income Statement 2021-01-01, valued at period ends

                     || 2021-01-01
=====================++============
 Revenues            ||
---------------------++------------
---------------------++------------
                     ||
=====================++============
 Expenses            ||
---------------------++------------
 expenses:A:Year2021 ||        $40
---------------------++------------
                     ||        $40
=====================++============
 Net:                ||       $-40
 ```

**Comment by dominikpaulus** [comment]
@apauley : What you are seeing is not the *transaction* that is filtered out, but the balance change on the expenses:A:Year2020-account that incurred by changing the price of "COM". It becomes clear if you change the prices in a way that P(COM, 2021) - P(COM, 2020) != P(COM, 2020):

```
$ echo 'P 2020-01-01 COM $15
P 2021-01-01 COM $50

2020-01-01 Trx 1
    assets:A
    expenses:A:Year2020     COM 1

2021-01-01 Trx 2
    assets:A
    expenses:A:Year2021     COM 2' | hledger -f - is --begin 2021-01-01 --exchange=$ --daily
Income Statement 2021-01-01

                     || 2021-01-01 
=====================++============
 Revenues            ||            
---------------------++------------
---------------------++------------
                     ||            
=====================++============
 Expenses            ||            
---------------------++------------
 expenses:A:Year2020 ||        $35 
 expenses:A:Year2021 ||       $100 
---------------------++------------
                     ||       $135 
=====================++============
 Net:                ||      $-135 
```

Changing the price of 'COM' causes a $50-$15 = $35 balance change in the report to show up.

**Comment by simonmichael** [comment]
Ah, I assume those remaining discrepancy/ies are all to do with 1.20's alteration to valued multiperiod balance change reports. This is being discussed at #1353 (and #1428).

I realised there's another problem with the new price lookup code: #1443. (Saying the word "perfect" did the trick!)

Meanwhile, the hang/exponential slowdown bug reported here has been fixed, but still to do:
- finish adding a timeout/iteration limit to guard against other similar bugs, and document it
- release the fix
